### DISCUSSION TITLE
Generate src and include without amalgamation

### DISCUSSION THREAD

**Discussion by shongoku** [randombit/botan#4332] [Q&A]
Hi,
The question is easy and I hope the answer too.
I need to create the include path with all declarative files with their structure; and the src path with all cpp files.
The question is that I can't use the amalgamation command because I must change perhaps one or more files in the future. Besides, I prefer to compile the library with my tools.
Does there exist a parameter that I can do this?
Thanks in advance.

**Comment by reneme** [comment]
What you describe sounds like the typical use case for the amalgamation build. Except that you expect to require patches to the library in the future.

The amalgamation build essentially just copies all headers and source files into two large files. I would suggest you apply your patches to the source and header files as you find them in the repository _and only then_ generate the amalgamation files.

That way, you can maintain your custom patches and benefit from the amalgamation build via your toolchain.

I hope I understand your question correctly. 😊

**Comment by shongoku** [comment]
Thank you so much for your answer reneme,

The question is your answer is not good to me.
Working with a file with thousands of lines is difficult. Furthermore, my computer is awful.
I don't understand why configure.py doesn't have an option to build src and include only necessary files and then we can put those files inside our project.
Perhaps it exists this option or a trick.

Thank you again reneme.

**Comment by reneme** [comment]
Judging from [your latest answer](https://github.com/randombit/botan/discussions/4332#discussioncomment-10530554), I'm assuming you want to specifically select the functionality of the library that you require for your project.

Indeed there's a way to do that: When you enable only the modules that you require for your project, the amalgamation files will be created from those modules only. This can reduce the overall size of the amalgamation files significantly. Please see [the documentation on how to configure a build](https://botan.randombit.net/handbook/building.html#configuring-the-build) for all the details. As well as [the list of all modules the library provides](https://botan.randombit.net/doxygen/topics.html).

### Example

Lets assume that the only things your project requires is a fast implementation of SHA-256 and an adapter to the operating system's RNG. Also, I'm assuming that you target a reasonably recent ARM platform. You might configure the library with:

```bash
./configure.py --minimized-build --enable-modules=sha2_32_armv8,system_rng --amalgamation
```

This should generate a pair of `botan_all.h` and `botan_all.cpp` files with less than 15.000 lines of code containing just the requested functionality plus some basic functions that must always be part of a minimal build.

Use `./configure.py --list-modules` to get a list of all available modules

Typically, it requires a bit of experimentation to reach a useful list of modules for your needs. Note that, as shown above, some algorithms might provide modules with platform-specific implementations that must be enabled explicitly. For instance, `--enable-modules=aes` will get you a basic software implementation of the AES and `--enable-modules=aes_armv8` yields a faster implementation based on ARM's crypto extensions.
